prototypes/ScheduledMerges.hs

```diff
@@ -220,11 +267,14 @@ invariant = go 1
     -- conditions are more complicated.
     expectedMergingRunLengths :: Int -> IncomingRun s -> MergingRunState
                               -> [Level s] -> ST s ()
-    expectedMergingRunLengths ln mr mrs ls =
+    expectedMergingRunLengths ln ir mrs ls =
       case mergePolicyForLevel ln ls of
-        MergePolicyLevelling ->
-          assert (mergeLastForLevel ls == MergeLastLevel) $
-          case (mr, mrs) of
+        MergePolicyLevelling -> do
+          case ir of
+            Merging _ (MergingRun mergeType _) ->
+              assertST (mergeType == MergeLevel (mergeLast ls))
+            _ -> pure ()
```

Comment:
True, that check is indeed redundant.

Reply:
I'm not sure it's worth enforcing these things statically. Maybe if we statically knew the exact type of merge, but in this case we only know that it's not a union merge. In other cases it can be union merge or last level merge, but not mid level.

We could also enforce a lot of other invariants statically, i.e. levelling and last level merges only being possible on the last level, but have so far chosen to check them dynamically instead to keep the types simple. Maybe it's worth giving it a try at some point, but this PR is already quite big.